java/connector-node/risingwave-sink-jdbc/src/main/java/com/risingwave/connector/JDBCSink.java

```diff
@@ -167,168 +127,250 @@ private static List<String> getPkColumnNames(
-            throw Status.FAILED_PRECONDITION
-                    .withDescription("unexpected op type: " + row.getOp())
-                    .asRuntimeException();
-        }
-        try {
-            var preparedStmt = insertPreparedStmt;
-            jdbcDialect.bindInsertIntoStatement(preparedStmt, conn, getTableSchema(), row);
-            preparedStmt.addBatch();
-            return preparedStmt;
-        } catch (SQLException e) {
-            throw io.grpc.Status.INTERNAL
-                    .withDescription(
-                            String.format(ERROR_REPORT_TEMPLATE, e.getSQLState(), e.getMessage()))
-                    .withCause(e)
-                    .asRuntimeException();
-        }
-    }
+    @Override
+    public void write(Iterable<SinkRow> rows) {
+        final int maxRetryCount = 4;
+        int retryCount = 0;
+        while (true) {
+            try {
+                // fill prepare statements with parameters
+                for (SinkRow row : rows) {
+                    if (row.getOp() == Data.Op.UPDATE_DELETE) {
+                        updateFlag = true;
+                        continue;
+                    }
+                    if (config.isUpsertSink()) {
+                        if (row.getOp() == Data.Op.DELETE) {
+                            jdbcStatements.prepareDelete(row);
+                        } else {
+                            jdbcStatements.prepareUpsert(row);
+                        }
+                    } else {
+                        jdbcStatements.prepareInsert(row);
+                    }
+                }
+                // Execute staging statements after all rows are prepared.
+                jdbcStatements.execute();
+                break;
+            } catch (SQLException e) {
+                LOG.error("Failed to execute JDBC statements, retried {} times", retryCount, e);
+                if (++retryCount > maxRetryCount) {
+                    throw Status.INTERNAL
+                            .withDescription(
+                                    "Failed to execute JDBC statements and exceeded max retry times")
+                            .withCause(e)
+                            .asRuntimeException();
+                }
 
-    private PreparedStatement prepareUpsertStatement(SinkRow row) {
-        try {
-            var preparedStmt = upsertPreparedStmt;
-            switch (row.getOp()) {
-                case INSERT:
-                    jdbcDialect.bindUpsertStatement(preparedStmt, conn, getTableSchema(), row);
-                    break;
-                case UPDATE_INSERT:
-                    if (!updateFlag) {
-                        throw Status.FAILED_PRECONDITION
-                                .withDescription("an UPDATE_DELETE should precede an UPDATE_INSERT")
-                                .asRuntimeException();
+                try {
+                    if (!conn.isValid(30)) { // 30 seconds timeout
```

Comment:
How is 30seconds timeout chosen? Will it be too long? I saw that the default isValid check timeout for JDBC conn is 3s.